### PR TITLE
[stable/3.0] backport #628

### PR DESCRIPTION
**Pull request by MaximilianMeister**
https://github.com/crowbar/crowbar-core/pull/628

### CODE REVIEW

**Comment by MaximilianMeister** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;xeN0LnPm&#x2F;95-5-s47p9-implement-api-for-preparing-nodes-for-upgrade-upgrade-step-2"> (5) S47P9 Implement API for preparing nodes for upgrade (upgrade step 2)</a>

File: crowbar_framework/app/models/crowbar_service.rb [hunk 1]
```diff
@@ -479,6 +479,25 @@ def prepare_nodes_for_os_upgrade
     upgrade_nodes_failed
   end
 
+  def revert_nodes_from_crowbar_upgrade
+    proposal = ProposalObject.find_proposal("crowbar", "default")
```
Comment (jsuchome) [inline crowbar_framework/app/models/crowbar_service.rb]: `proposal = Proposal.find_by(barclamp: "crowbar", name: "default")`

**Comment by jsuchome** [comment]
We also need the chef bits from https://github.com/crowbar/barclamp-crowbar/blob/release/tex/master/chef/cookbooks/crowbar/recipes/crowbar-upgrade.rb#L66

And also for the part we've just merged (https://github.com/crowbar/crowbar-core/pull/628)

**Comment by jsuchome** [comment]
3 . I think we also need to add more states where crowbar-upgrade can be applied, like it is here https://github.com/crowbar/barclamp-crowbar/blob/release/tex/master/chef/data_bags/crowbar/bc-template-crowbar.json#L37 . Otherwise the chef part is not executed when reverting from upgrade.

**Comment by jsuchome** [comment]
@MaximilianMeister so, I've put the stuff that's needed here (PR to master): https://github.com/crowbar/crowbar-core/pull/631

File: crowbar_framework/app/models/crowbar_service.rb (hunk 1 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/models/crowbar_service.rb]: right, the `ProposalObject` is long gone...

**Comment by MaximilianMeister** [comment]
> so, I've put the stuff that's needed here (PR to master): #631

thanks, so lets wait until it is merged and backported, and I'll just rebase this PR here on top afterwards

File: crowbar_framework/app/models/crowbar_service.rb (hunk 1 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/models/crowbar_service.rb]: done

fix for `master` is in https://github.com/crowbar/crowbar-core/pull/632

**Comment by aplanas** [comment]
Are the backported needed done?

**Comment by MaximilianMeister** [comment]
> Are the backported needed done?

@aplanas https://github.com/crowbar/crowbar-core/pull/634 needs to be merged still

**Comment by nicolasbock** [comment]
+1

**Comment by jsuchome** [comment]
+1

**Pull request merged**